tomviz/RotateAlignWidget.cxx

```diff
@@ -0,0 +1,503 @@
+    this->Internals->mainSliceMapper->Update();
+  }
+  updateWidgets();
+}
+
+void RotateAlignWidget::onProjectionNumberChanged(int newVal)
+{
+  vtkTrivialProducer *t =
+      vtkTrivialProducer::SafeDownCast(this->Internals->Source->producer()->GetClientSideObject());
+  if (!t)
+  {
+    std::cout << "Failed to get trivail producer" << std::endl;
+    return;
+  }
+  vtkImageData *imageData = vtkImageData::SafeDownCast(t->GetOutputDataObject(0));
+  int extent[6];
+  imageData->GetExtent(extent);
+  this->Internals->mainSliceMapper->SetSliceNumber(newVal + extent[4]);
+  this->Internals->mainSliceMapper->Update();
+  this->Internals->Ui.sliceView->update();
+}
+
+void RotateAlignWidget::onRotationAxisChanged()
+{
+  this->Internals->moveRotationAxisLine();
+}
+
+void RotateAlignWidget::onRotationAngleChanged()
+{
+  this->Internals->moveRotationAxisLine();
+}
+
+void RotateAlignWidget::onReconSliceChanged(int)
+{
+  QSpinBox *sb = qobject_cast<QSpinBox*>(this->sender());
+  this->Internals->updateSliceLines();
+  this->Internals->Ui.sliceView->update();
+  if (sb == this->Internals->Ui.spinBox_1)
+  {
+    this->Internals->updateReconSlice(0);
+//    this->Internals->reconSliceMapper[0]->SetSliceNumber(newSlice);
+    this->Internals->reconSliceMapper[0]->Update();
+    this->Internals->Ui.sliceView_1->update();
+  }
+  else if (sb == this->Internals->Ui.spinBox_2)
+  {
+    this->Internals->updateReconSlice(1);
+//    this->Internals->reconSliceMapper[1]->SetSliceNumber(newSlice);
+    this->Internals->reconSliceMapper[1]->Update();
+    this->Internals->Ui.sliceView_2->update();
+  }
+  else // if (sb == this->Internals->Ui.spinBox_3)
+  {
+    this->Internals->updateReconSlice(2);
+//    this->Internals->reconSliceMapper[2]->SetSliceNumber(newSlice);
+    this->Internals->reconSliceMapper[2]->Update();
+    this->Internals->Ui.sliceView_3->update();
+  }
+}
+
+void RotateAlignWidget::updateWidgets()
+{
+  this->Internals->Ui.sliceView->update();
+  this->Internals->Ui.sliceView_1->update();
+  this->Internals->Ui.sliceView_2->update();
+  this->Internals->Ui.sliceView_3->update();
+}
+
+void RotateAlignWidget::onFinalReconButtonPressed()
+{
+  DataSource *source = this->Internals->Source;
+  if (!source)
+  {
+    return;
+  }
+  vtkTrivialProducer *t =
+      vtkTrivialProducer::SafeDownCast(source->producer()->GetClientSideObject());
+  vtkImageData *imageData = vtkImageData::SafeDownCast(t->GetOutputDataObject(0));
+  DataSource* output = source->clone(true,true);
+  QString name = output->producer()->GetAnnotation("tomviz.Label");
+  name = "Rotation_Aligned_" + name;
+  output->producer()->SetAnnotation("tomviz.Label", name.toAscii().data());
+  t = vtkTrivialProducer::SafeDownCast(output->producer()->GetClientSideObject());
+  vtkImageData *recon = vtkImageData::SafeDownCast(t->GetOutputDataObject(0));
+
+  TomographyReconstruction::weightedBackProjection3(imageData, recon);
```

Comment:
No need to perform a 3D reconstruction. This UI is only for generating aligned tilt series. Users may want to use other reconstruction methods.